### ISSUE TITLE
seeing lots of metron agent panics after upgrading from v231 to v236

### ISSUE THREAD

**Issue by rwwohl** [cloudfoundry/loggregator#127]
Noticed that our metron agents having many panics throughout the day at random times with following error:

panic: No enabled dopplers available, check your manifest to make sure you have dopplers listening for the following protocols [udp]

After looking into it noticed that many times it correlated to etcd re-elections which were happening frequently due to slow disk i/o on etcd servers.  I adjusted the election timeout a bit and that has helped alot but still seeing occasionally panics.  Metron agents seem very fragile now if they have any interruptions with etcd or if the dopplers fail to update their entries for some reason.

**Comment by nelsam** [comment]
@rwwohl: Metron is most likely getting events from `etcd` with no dopplers (all doppler endpoints have disappeared or timed out), which causes it to panic.  Based on what you said, it sounds like this is related to dopplers being unable to refresh their `TTL` on their endpoints in etcd.

Can you check your `etcd` nodes for any clock drift?  We've had problems with etcd's method of dealing with TTL values (their logic assumes that nodes have zero clock drift), which caused metron to go a bit insane.

The reason we updated metron to panic in this scenario is that it was silently dropping messages so long as it had no doppler(s) to talk to.  If it receives an event from etcd, it needs to respect the addresses available in that event, and if it loses all dopplers, it's going to be dropping a lot of data.  That seems like a valid situation to panic in.  Monit will restart it, but the crashed metron will alert you that there's a serious problem in your environment.

**Comment by rwwohl** [comment]
Actually, clock drift was the first thing I suspected because I have seen that before with the TTL's.  Thats not the case on this one.  We have 4 environments (2 in aws and 2 in vsphere) that are all displaying this behavior and I cannot pin down the root cause.  Can log volume the dopplers are handling cause issues with them updating their TTL's in etcd on time?  Just trying to hunt down other things that could be causing this.  Seems unlikely to be network since there are so many environments having the same issue.

**Comment by iplay88keys** [comment]
As far as we are aware of, log volume shouldn't be a factor in Doppler updating its `TTL`.
In order to figure out what's going on, how many `etcd(s)` and `doppler(s)` do you have in those environments?

Additionally, could you take a look at the logs for `etcd` and `Doppler` to see if there is anything that stands out?

**Comment by rwwohl** [comment]
We have 2 dopplers and 3 etcd servers in the environments.  Nothing stands out in the doppler logs.  The restarts for the metrons always coorespond to etcd logs that either indicate an election or we see entries like below in all 3 etcd servers:

2016/06/15 21:20:21 etcdhttp: got unexpected response error (etcdserver: request timed out)
2016/06/15 21:20:22 etcdhttp: got unexpected response error (etcdserver: request timed out)
2016/06/15 21:20:22 etcdhttp: got unexpected response error (etcdserver: request timed out)
2016/06/15 21:20:22 etcdhttp: got unexpected response error (etcdserver: request timed out)
2016/06/15 21:20:22 etcdhttp: got unexpected response error (etcdserver: request timed out)

As mentioned, we used to see this many times a day with constant elections in etcd.  I adjusted the election timeout up to 2500 and that has reduced elections quite a bit(from several times a day to maybe once a day), but we still see the above type of logging.  I'm leaning toward disk i/o since I see how wait times on the etcd server disks but not sure if that would correlate with messages above.
